### PR TITLE
Optimize the implementation of readArray and readByteBuffer (Zero copy POC)

### PR DESCRIPTION
**Pull request by luozhenyu**
### Type of change

- Refactoring

#### readArray
1. Remove bounds checking codes and the bounds checking is now doned by `ByteBufUtil.getBytes`.
2. `ByteBufUtil.getBytes` has better performance(0~10% faster depending on JDK version and bytebuf type) than simply allocating bytes and copying. For direct buffer, it uses `PlatformDependent.allocateUninitializedArray` which does not require zeroing elimination. For heap buffer, `ByteBufUtil.getBytes` has fewer bounds checking overhead.

#### readByteBuffer
1. Avoid copying the internal nio buffer. (3x faster when the buffer size is 16 bytes, 22x faster when the buffer size is 1024 bytes)

### Additional Context

We are doing performance testing and have found some codes that could be improved. The most significant one is fixed by #1192 

### Checklist

- [x] Write tests
- [x] Make sure all tests pass
- [x] Review performance test results. Ensure that any degradations to performance numbers are understood and justified.
- [ ] Make sure all Sonarcloud warnings are addressed or are justifiably ignored.
- [ ] Update documentation
- [ ] Reference relevant issue(s) and close them after merging
- [ ] For user facing changes, update CHANGELOG.md (remember to include changes affecting the API of the test artefacts too).

### CODE REVIEW

**Comment by franz1981** [comment]
Hold on: allocated not initialized have some limits sadly,  because Netty 4.1 cannot use MethodHandles (while Netty 4.2, yep 👍) and that's why on Netty there are limits when is applied (it is uswd beyond configueable limits); the reflective cost rwsuce the application for small arrays...

**Comment by k-wall** [comment]
Thanks for the feedback @franz1981, having your Netty expertise is really appreciated.   Your whole comment wasn't super easy to parse, but I'm getting that when Netty 4.2 lands we'll be able to make a performance improvement safely.  Could you tell us more about what we'll need to do and how it helps?

**Comment by franz1981** [comment]
Sorry @k-wall writing from phone is never a good idea :/

let me rephrase it a bit_
- with 4.2 some Netty methods which were using reflection, will use MathodHandle, improving their performance ten-fold (in some scenario), see my comment https://github.com/netty/netty/pull/14032#issuecomment-2094285788 and the change at https://github.com/netty/netty/pull/14032/files#diff-e500fe6cdcd642c7929cdbb2bc118760d09f400093ff34d63d8272d35f80cf8bR293
- one of such methods is `Unsafe::allocateUninitializedArray` which can speed up A LOT byte[] allocations within Netty, but in the 4.1 version contains the mentioned `io.netty.uninitializedArrayAllocationThreshold` limit (removed in 4.2 because of my finding re its performance with MethodHandle)

This translates in: beware `PlatformDependent.allocateUninitializedArray` because it kicks in for > 1024 bytes, in Netty 4.1

File: kroxylicious-runtime/src/main/java/io/kroxylicious/proxy/internal/codec/ByteBufAccessorImpl.java [hunk 1]
```diff
@@ -138,13 +133,9 @@ public double readDouble() {
...
     @Override
     public byte[] readArray(int size) {
-        int remaining = buf.readableBytes();
-        if (size > remaining) {
-            throw illegalReadException(size, remaining);
-        }
-        byte[] dst = new byte[size];
-        buf.readBytes(dst, 0, size);
-        return dst;
+        byte[] array = ByteBufUtil.getBytes(buf, buf.readerIndex(), size, false);
```
Comment (k-wall) [inline kroxylicious-runtime/src/main/java/io/kroxylicious/proxy/internal/codec/ByteBufAccessorImpl.java:136]: Doesn't this change semantics? It is safe.  Previously the code returned a copy of the bytes, whereas now (copy=false) it is potential returning the live bytes.    There's nothing in the contact io.kroxylicious.proxy.frame.ByteBufAccessor#readArray or org.apache.kafka.common.protocol.Readable#readArray that actually states what behaviour should be.   I notice that the ResponseData objects propagate the reference.  We should be cautious before making a change.

AFAIK, I think the change might be safe and I don't think it actually changes anything for the Filter Author.  If we decide to go this away, we should javadoc ByteBufAccessor accordingly.

File: kroxylicious-runtime/src/main/java/io/kroxylicious/proxy/internal/codec/ByteBufAccessorImpl.java [hunk 2]
```diff
@@ -154,11 +145,9 @@ public int readUnsignedVarint() {
 
     @Override
     public ByteBuffer readByteBuffer(int length) {
-        // TODO use buf.nioBufferCount() and buf.nioBuffers() to avoid the copy if possible
-        ByteBuffer wrap = ByteBuffer.wrap(ByteBufUtil.getBytes(buf, buf.readerIndex(), length, false));
+        ByteBuffer byteBuffer = buf.internalNioBuffer(buf.readerIndex(), length);
```
Comment (k-wall) [inline kroxylicious-runtime/src/main/java/io/kroxylicious/proxy/internal/codec/ByteBufAccessorImpl.java:148]: Couple of comments here:

1. `io.netty.buffer.ByteBuf#internalNioBuffer` is documented as internal use only.  I'm circumspect about disregarding that notice.

2. There's a change to semantics.  Previously the returned `ByteBuffer` belonged to the caller and the bytes contained were a copy.   The code is now exposing the underlying ByteBuffer object so if the caller calls say `ByteBuffer#mark`, might affect Netty's own operation.    If we did decide to do this, we should properly return a `ByteBuffer#slice` so the marks etc are independent.

3. I've got the same concerns about returning the live bytes rather than a copy.  Like I said above AFAIK this might be okay.

**Comment by k-wall** [comment]
@luozhenyu thanks for the contribution and taking the time to benchmark. I think there's a couple of things for us all to discuss to help us decide the right path.

**Comment by k-wall** [comment]
@strimzi-ci run perf

**Comment by k-wall** [comment]
Hello @luozhenyu,  I think it might help reviewers if you raise your Zero copy work (2nd commit) as a separate PR.  It seems to be going quite a bit deeper than the first.  In general, when tackling a larger problem you want to start with an issue describing the problem you see (with reproduction) and talk about possible solutions, then let PR(s) follow from that.  Keeping PRs narrowly focused helps reviewers do their job.  Thank you again for the contributions.

**Comment by k-wall** [comment]
@luozhenyu would you be able to respond to my comment above?

**Comment by franvila** [comment]
There is no reply from the author, so closing it.

**Comment by showuon** [comment]
@luozhenyu , thanks for your contribution!
For this PR, you did:

```
readArray
   1. Remove bounds checking codes and the bounds checking is now doned by ByteBufUtil.getBytes.
   2. ByteBufUtil.getBytes has better performance(0~10% faster depending on JDK version and bytebuf type) than simply allocating bytes and copying. For direct buffer, it uses PlatformDependent.allocateUninitializedArray which does not require zeroing elimination. For heap buffer, ByteBufUtil.getBytes has fewer bounds checking overhead.

readByteBuffer
   3. Avoid copying the internal nio buffer. (3x faster when the buffer size is 16 bytes, 22x faster when the buffer size is 1024 bytes)
```
For (1), we agree that's a good refactor. Welcome to open a separate PR for it.
For (2) and (3), since it will go quite deeper, and has many changes, maybe it'd be better you start an [issue](https://github.com/kroxylicious/kroxylicious/issues) first to provide more context about the issue. After we reach consensus, we can open PRs for it then. WDYT?

Again, thanks for improving kroxylicious!
